spring-rabbit-test/src/main/java/org/springframework/amqp/rabbit/test/RabbitListenerTestHarness.java

```diff
@@ -0,0 +1,113 @@
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.amqp.rabbit.annotation.RabbitListener;
+import org.springframework.amqp.rabbit.annotation.RabbitListenerAnnotationBeanPostProcessor;
+import org.springframework.amqp.rabbit.listener.MethodRabbitListenerEndpoint;
+import org.springframework.aop.framework.ProxyFactoryBean;
+import org.springframework.util.StringUtils;
+
+/**
+ * BeanPostProcessor extending {@link RabbitListenerAnnotationBeanPostProcessor}.
+ * Wraps the listener bean in a CGLIB proxy with an advice to capture the arguments
+ * and result (if any) in a blocking queue. Test cases can access the results
+ * by autowiring the test harness into test cases.
+ *
+ * @author Gary Russell
+ * @since 1.6
+ *
+ */
+public class RabbitListenerTestHarness extends RabbitListenerAnnotationBeanPostProcessor {
+
+	private final Log logger = LogFactory.getLog(this.getClass());
+
+	private final Map<String, CaptureAdvice> listenerCapture = new HashMap<String, CaptureAdvice>();
+
+	@Override
+	protected void processListener(MethodRabbitListenerEndpoint endpoint, RabbitListener rabbitListener, Object bean,
+			Object adminTarget, String beanName) {
+		String id = rabbitListener.id();
+		if (StringUtils.hasText(id)) {
+			try {
+				ProxyFactoryBean pfb = new ProxyFactoryBean();
+				pfb.setProxyTargetClass(true);
+				pfb.setTarget(bean);
+				CaptureAdvice advice = new CaptureAdvice();
+				pfb.addAdvice(advice);
+				Object advised = pfb.getObject();
+					this.listenerCapture.put(id, advice);
+					super.processListener(endpoint, rabbitListener, advised, adminTarget, beanName);
+					return;
+			}
+			catch (Exception e) {
+				logger.error("Failed to proxy @RabbitListener with id: " + id);
+			}
+		}
+		else {
+			logger.info("The test harness can only proxy @RabbitListeners with an 'id' attribute");
+		}
+		super.processListener(endpoint, rabbitListener, bean, adminTarget, beanName);
+	}
+
+	public BlockingQueue<Object[]> getArgumentQueueFor(String id) {
+		CaptureAdvice advice = this.listenerCapture.get(id);
+		if (advice != null) {
+			return advice.arguments;
+		}
+		return null;
+	}
+
+	public BlockingQueue<Object> getResultQueueFor(String id) {
+		CaptureAdvice advice = this.listenerCapture.get(id);
+		if (advice != null) {
+			return advice.returns;
+		}
+		return null;
+	}
+
+	private static final class CaptureAdvice implements MethodInterceptor {
+
+		private final BlockingQueue<Object[]> arguments = new LinkedBlockingQueue<Object[]>();
+
+		private final BlockingQueue<Object> returns = new LinkedBlockingQueue<Object>();
+
+		@Override
+		public Object invoke(MethodInvocation invocation) throws Throwable {
+			Object[] args = invocation.getArguments();
+			if (args.length > 0) {
+				arguments.add(args);
+			}
+			Object result = invocation.proceed();
+			if (args.length > 0 && result != null) {
+				this.returns.put(result);
+			}
+			return result;
```

Comment:
Continuing the `ArgumentsAndResult` discussion, I think here we should be robust and even don't expose arguments until the `invocation.proceed()` returns.
Anyway there should be some hook to expect an `Exception`- `try...catch` around `invocation.proceed()`.